### PR TITLE
Add swipe action on mobile to switch between dashboard views

### PR DESCRIPTION
**Pull request by LucasTor**
## Proposed change

As an iPhone user, I'm really used to navigating with gestures, so this PR adds the ability to swipe between views in a dashboard
It will only swipe if the swipe length is at least 30% of the view width and lasts at max 200ms, to prevent false positives

WhatsApp-Video-2022-11-07-at-17 04 23

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

No additional configuration needed

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Review by steverep — COMMENTED** [review]
Just some general comments and questions:

1. Where did the 30% and 200 ms come from?  Is that a documented best practice somewhere and consistent with OS behavior?
2. @piitaya introduced [hammer.js](https://hammerjs.github.io) for handling drag and drop, so using that library for swipe as well is probably the better approach than hard coding in HA
3. For the least confusing UX, there should be a plan to also support swipe in other major tabbed areas, such as developer tools and settings pages.

**Comment by LucasTor** [comment]
@steverep refactored the code to use Hammer.js, thanks for the tip :)

The values for the timing and amount came by testing, just like I had to do with Hammer to get it to work properly, I tested on both Android and iOS devices

I also created a small "abstraction" to setup the lateral swipe where you pass in the callbacks for handling that in the component and it doe the rest, I implemented them in the Developer tools as well as the dashboard, do you think I should implement it in all the other places as well? This would probably take a while as I don't usually have that much time to spare :/

Thanks for the review!

File: src/util/swipe.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import "hammerjs";
+
+export const setupHorizontalSwipe = (backwardsCallback, forwardsCallback) => {
+  const isMobile = /iPhone|iPad|iPod|Android/i.test(navigator.userAgent);
+
+  if (!isMobile) return () => {};
```
Comment (steverep) [inline src/util/swipe.ts:6]: Why restrict only to mobile?

File: src/util/swipe.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+  const mc = new Hammer.Manager(document.body);
+  mc.add(
+    new Hammer.Swipe({
+      threshold: 120,
+      velocity: 1,
```
Comment (steverep) [inline src/util/swipe.ts:12]: These parameters are much larger than the defaults so I'm just concerned it requires a swipe to be faster and longer than is typical.  That said, I cannot find any documented guidelines.  [Apple's swipe recognizer](https://developer.apple.com/documentation/uikit/uiswipegesturerecognizer) describes fast and slow possibilities but has no useful details.

@matthiasdebaat do you know where there are any UX guidelines on the gritty specifics of gestures?

**Review by steverep — COMMENTED** [review]
To work on the settings pages, you'd only have to modify `hass-tabs-subpage`.

**Comment by LucasTor** [comment]
@steverep Just added the swipe function to the hass-tabs-subpage, now it works on the settings page as well, thanks for that.
I guess now we can wait on @matthiasdebaat to give his considerations on the swipe action (velocity, length...), and maybe tell us whether it would be a good thing to keep it on desktops as well

**Comment by steverep** [comment]
@LucasTor I haven't had a chance to pull this branch and play with it yet, but I was just wondering what the behavior is when a more info or other dialog is opened.  Does it respect the inertness of the background and ignore swipes?

**Comment by LucasTor** [comment]
@steverep I wasn't sure so I had to test it, apparently, if a dialog or something else is open in front of the swipe target, it respects the inertness

**Comment by steverep** [comment]
> @steverep I wasn't sure so I had to test it, apparently, if a dialog or something else is open in front of the swipe target, it respects the inertness

Was the dialog you tested always full screen or is it respected even on the backdrop?  Might need a larger tablet or desktop to test that.

**Comment by LucasTor** [comment]
> > @steverep I wasn't sure so I had to test it, apparently, if a dialog or something else is open in front of the swipe target, it respects the inertness
> 
> Was the dialog you tested always full screen or is it respected even on the backdrop? Might need a larger tablet or desktop to test that.

You are right, in bigger screens where the dialog is full screen you can still swipe in the backdrop, though changing the Hammer.Swipe target element to this.shadowRoot seems to have solved the problem, sadly my tablet is not big enough to test it (8in only), so I had to test it in Chrome's devtools

**Comment by matthiasdebaat** [comment]
Sorry for the hold up, I'm back from my paternity leave. 

I like adding gestures. One much requested gesture is swiping back to previous page. Would these gestures work together? 

> I guess now we can wait on @matthiasdebaat to give his considerations on the swipe action (velocity, length...), and maybe tell us whether it would be a good thing to keep it on desktops as well

Material design describes [these](https://m2.material.io/design/interaction/gestures.html#types-of-gestures) types of gestures and [speed](https://m2.material.io/design/motion/speed.html#duration) motion. Couldn't find any documentation about velocity and length. Do you know what the defaults are?

Does Home Assistant knows when a device got a touchscreen? For example, I've seen some people using big tablets for their dashboards. Why don't we want to have swipe actions on desktop?

**Comment by bramkragten** [comment]
> Does Home Assistant knows when a device got a touchscreen? For example, I've seen some people using big tablets for their dashboards. Why don't we want to have swipe actions on desktop?

Yes we can see this

One big issues with these kind of actions is that we have sliders and other ui elements that also use swipe gestures, how do we prevent these gestures to collide?

**Comment by steverep** [comment]
> One big issues with these kind of actions is that we have sliders and other ui elements that also use swipe gestures, how do we prevent these gestures to collide?

We do?  😮  Where are swipes supported now?



> I like adding gestures. One much requested gesture is swiping back to previous page. Would these gestures work together?

I think we'd have to pick one or the other where both tabs and a back button exist.  Having swipes control the tabs makes more sense to me. 🤷‍♂️

**Comment by matthiasdebaat** [comment]
> One big issues with these kind of actions is that we have sliders and other ui elements that also use swipe gestures, how do we prevent these gestures to collide?

iOS and Android have a solution for this. UI elements go for swipe gestures, if you swipe over a slider it slides.

> I think we'd have to pick one or the other where both tabs and a back button exist. Having swipes control the tabs makes more sense to me. 🤷‍♂️

Default mobile gestures depends on where a users starts swiping. On the side is a swipe back and in the middel switches tabs. Codewise I have no idea if this can be done. 

If we have to pick one I would suggest the page back, this is a more often interaction and much requested feature.

**Comment by steverep** [comment]
> Default mobile gestures depends on where a users starts swiping. On the side is a swipe back and in the middel switches tabs. Codewise I have no idea if this can be done.

Yes, more complicated and we'd need to know what the typical OS cutoff point between "side" and "middle" are, but can be done.

**Comment by LucasTor** [comment]
First of all, congrats @matthiasdebaat :)

> Material design describes [these](https://m2.material.io/design/interaction/gestures.html#types-of-gestures) types of gestures and [speed](https://m2.material.io/design/motion/speed.html#duration) motion. Couldn't find any documentation about velocity and length. Do you know what the defaults are?

I'll have a look at those guidelines and see what I can figure out based on that, the default values for hammer are way smaller that what we have rn, but it is completely unusable because of how sensitive it gets to swipes

> iOS and Android have a solution for this. UI elements go for swipe gestures, if you swipe over a slider it slides.

If I'm not mistaken, Hammer already deals with that, I'm currently traveling right now and cannot test it, but I had a slider on my test instance and if I recall correctly I could use it normally without the swipe actions affecting it

> Default mobile gestures depends on where a users starts swiping. On the side is a swipe back and in the middel switches tabs. Codewise I have no idea if this can be done.

Codewise, we can totally do that, I don't see a problem, the problem I see with that is that currently when on mobile, if you swipe from the left, you open the sidebar, would we disable that functionality in favor of the back action?

**Comment by matthiasdebaat** [comment]
> First of all, congrats @matthiasdebaat :)

Thanks! ❤️

> Codewise, we can totally do that, I don't see a problem, the problem I see with that is that currently when on mobile, if you swipe from the left, you open the sidebar, would we disable that functionality in favor of the back action?

Cool! An often used interaction design is when a chevron back is shown at the top left you can also swipe back. In our case this means we have both swipe gestures: hamburger menu icon is open sidebar, chevron back icon is page back.

**Comment by steverep** [comment]
Yes @matthiasdebaat, meant to say congrats as well!

> currently when on mobile, if you swipe from the left, you open the sidebar, would we disable that functionality in favor of the back action?

Really? 😮  That doesn't seem to work for me on iOS.


In any case, does the back button and sidebar button ever co-exist on mobile?

**Comment by LucasTor** [comment]
> Really? 😮 That doesn't seem to work for me on iOS.

Yes! Just tested on iOS 16 on my iPhone 13 and it works fine
(had to record it on my PC bc I did not find a way to place a cursor on iOS recordings)
ezgif-2-a95b5926a8

> In any case, does the back button and sidebar button ever co-exist on mobile?

Likely not, I think we have either the back button or the sidebar button on the top left of the screen, probably Matthias can confirm that for us

**Comment by matthiasdebaat** [comment]
> Likely not, I think we have either the back button or the sidebar button on the top left of the screen, probably Matthias can confirm that for us

On smaller screen sizes it's only menu or back button.

**Comment by Daneish** [comment]
> One much requested gesture is swiping back to previous page.

Just a quick note, as the feature doesn’t seem to be well known, but in the iOS app, a two finger swipe from left to right is a back gesture. I’m not sure if browsers can register multi-touch gestures, (and obviously won’t work with a mouse) but it might be useful to try to keep experiences similar across devices.

Also, congratulations Matthias!

**Comment by bramkragten** [comment]
How will it handle the combination with this card? <https://github.com/bramkragten/swipe-card> 😜

> iOS and Android have a solution for this. UI elements go for swipe gestures, if you swipe over a slider it slides.

Not sure I understand what you mean... if you have a slider, like this one for example <https://github.com/home-assistant/frontend/pull/14446>, how would the difference between sliding the slider and sliding to a different dashboard be made?

You mean that if an element handles the swipe action, the parent (switch views) would not handle it?

**Comment by matthiasdebaat** [comment]
> Just a quick note, as the feature doesn’t seem to be well known, but in the iOS app, a two finger swipe from left to right is a back gesture. I’m not sure if browsers can register multi-touch gestures, (and obviously won’t work with a mouse) but it might be useful to try to keep experiences similar across devices.

Ha, didn't know that one. Thanks! Think we should replace it with the more common single finger swipe back.

> Not sure I understand what you mean... if you have a slider, like this one for example #14446, how would the difference between sliding the slider and sliding to a different dashboard be made?
> 
> You mean that if an element handles the swipe action, the parent (switch views) would not handle it?

Yes

**Comment by dshokouhi** [comment]
We should make sure to test this change against Androids swipe navigation gesture system. I am not sure about iOS but on Android users have the ability to control how much drag space they can do with the gestures, from left and right edge independently. In the GIF above android will probably register the back swipe depending on how fast the user swipes. I can only get the menu to pop-up if i swipe slowly for example.

Currently there are 2 open frontend issues (that happen in browser and app) that might help with some considerations

https://github.com/home-assistant/frontend/issues/14468
https://github.com/home-assistant/frontend/issues/10348

With that said should this feature be something that a user can enable/disable so that it does not interfere with normal navigation? I imagine if a user tries to swipe back in the app and then gets a tab moved that may be unexpected and undesired.

**Comment by matthiasdebaat** [comment]
> I imagine if a user tries to swipe back in the app and then gets a tab moved that may be unexpected and undesired.

I agree, we should avoid that. Would love to follow the common swipe gestures. If we can make that happen without adding extra settings would be nice.

**Comment by LasseRosenow** [comment]
Sorry this is a bit out of context, but I was also thinking maybe the better solution is to convert the tabs on mobile to a bottom navigation bar just like some pages in settings already do I think.

Then we also would not need to care about swipe gestures et al

**Comment by matthiasdebaat** [comment]
> Sorry this is a bit out of context, but I was also thinking maybe the better solution is to convert the tabs on mobile to a bottom navigation bar just like some pages in settings already do I think.
> 
> Then we also would not need to care about swipe gestures et al

Our tabbars have a maximum of 4 items, that wouldn't fit for dashboards with more tabs

**Comment by steverep** [comment]
IMHO, this is still a worthwhile feature to pursue.

@bramkragten do you see a way forward here?

**Comment by dnestico** [comment]
> Sorry this is a bit out of context, but I was also thinking maybe the better solution is to convert the tabs on mobile to a bottom navigation bar just like some pages in settings already do I think.
> 

I know this is pretty old now but I would love the option to enable a tab bar on mobile, yea most people have more than 4 tabs but I can definitely make it work on mine so would be a nice option.

File: src/util/swipe.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import "hammerjs";
```
Comment (bramkragten) [inline src/util/swipe.ts:1]: we use `@egjs/hammerjs`

File: src/util/swipe.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+export const setupHorizontalSwipe = (
+  backwardsCallback,
+  forwardsCallback,
+  element
```
Comment (bramkragten) [inline src/util/swipe.ts:6]: Please add typing

**Comment by bramkragten** [comment]
Was playing around with [view transitions](https://developer.chrome.com/docs/web-platform/view-transitions/) a bit, would be pretty cool here:


https://github.com/home-assistant/frontend/assets/5662298/e0e91481-444f-4941-b97b-212eeeba1a07

**Comment by bramkragten** [comment]
When panning a map, it will switch tabs, we should prevent that

**Comment by LucasTor** [comment]
I guess we could revisit this one, as it's still something I would like to have, maybe it could also be developed as a HACS frontend integration, if that would make it any easier to be used

**Comment by dshokouhi** [comment]
> maybe it could also be developed as a HACS frontend integration

you mean like this

https://github.com/zanna-37/hass-swipe-navigation

**Pull request closed without merging**